adal/tests/Test.ADAL.NET.Integration/DeviceProfileTests.cs

```diff
@@ -231,5 +235,31 @@ public void NegativeDeviceCodeTest()
 
             Assert.AreEqual(0, AdalHttpMessageHandlerFactory.MockHandlersCount());
         }
+
+        [TestMethod]
+        public async Task NegativeDeviceCodeTimeoutTest_WithZeroRetries()
+        {
+            MockHttpMessageHandler mockMessageHandler = new MockHttpMessageHandler(TestConstants.DefaultAuthorityHomeTenant)
+            {
+                Method = HttpMethod.Get,
+                Url = TestConstants.DefaultAuthorityHomeTenant + "oauth2/devicecode",
+                ResponseMessage = MockHelpers.CreateSuccessDeviceCodeResponseMessage(
+                    expirationTimeInSeconds: 0,
+                    retryInternvalInSeconds: 1)
+            };
+
+            AdalHttpMessageHandlerFactory.AddMockHandler(mockMessageHandler);
+
+            TokenCache cache = new TokenCache();
+            context = new AuthenticationContext(TestConstants.DefaultAuthorityHomeTenant, cache);
+            DeviceCodeResult dcr = await context.AcquireDeviceCodeAsync("some resource", "some authority");
+
+            Assert.IsNotNull(dcr);
+            AuthenticationResult result;
+            AdalServiceException ex = AssertException.TaskThrows<AdalServiceException>(async () => result = await context.AcquireTokenByDeviceCodeAsync(dcr));
+            Assert.IsTrue(ex.Message.Contains("Verification code expired"));
```

Comment:
I know that historically this code base uses `Assert.IsTrue(ex.Message.Contains("blah blah blah"))` a lot. But, at least for newly added test cases, do we consider to switch to a different style as `Assert.AreEqual(AdalErrorEx.DeviceCodeAuthorizationCodeExpired, ex.ErrorCode)`?

Rationale: When error code and error message are both available, an unspoken assumption is that the messages are considered as documentation therefore could potentially be changed in future (in order to fix typo, or rewrite for a better explanation, etc.), but the error code would be considered as part of the contract and remains the same (unless the designer would introduce a breaking change).

Reply:
Yes, great point!